pallets/identity/src/lib.rs

```diff
@@ -2226,42 +2226,65 @@ impl<T: Trait> CheckAccountCallPermissions<T::AccountId> for Module<T> {
+                    // In this case, we check the permissions of `who` in `target_did` as it is a direct call
+                    if who != &target_did_record.primary_key {
+                        // `who` can be a secondary key.
+                        //
+                        // DIDs with frozen secondary keys (aka frozen DIDs) are not permitted to call
+                        // extrinsics.
+                        if Self::is_did_frozen(&target_did) {
+                            // `target_did` has its secondary keys frozen.
+                            return None;
+                        }
+                        return target_did_record
+                            .secondary_keys
+                            .into_iter()
+                            .find(|sk| sk.signer == Signatory::Account(who.clone()))
+                            .filter(|sk| sk.has_extrinsic_permission(pallet_name, function_name))
+                            .map(|sk| AccountCallPermissionsData {
+                                primary_did: target_did,
+                                secondary_key: Some(sk),
+                            });
+                    }
+                } else {
+                    // In this case, we check the permissions of `who` in `target_did` as it is a direct call
+                    // DIDs with frozen secondary keys (aka frozen DIDs) are not permitted to call
+                    // extrinsics.
+                    if Self::is_did_frozen(&target_did) {
+                        // `target_did` has its secondary keys frozen.
+                        return None;
+                    }
+                    return target_did_record
+                        .secondary_keys
+                        .into_iter()
+                        .find(|sk| sk.signer.clone() == Signatory::Identity(key_did))
+                        .filter(|sk| sk.has_extrinsic_permission(pallet_name, function_name))
+                        .map(|sk| AccountCallPermissionsData {
+                            primary_did: target_did,
+                            secondary_key: Some(sk),
+                        });
                 }
-                let maybe_current_did_signer =
-                    Context::current_identity::<Self>().map(|did| Signatory::Identity(did));
-                return did_record
-                    .secondary_keys
-                    .into_iter()
-                    .find(|sk| {
-                        sk.signer == Signatory::Account(who.clone())
-                            || Some(sk.signer.clone()) == maybe_current_did_signer
-                    })
-                    .filter(|sk| sk.has_extrinsic_permission(pallet_name, function_name))
-                    .map(|sk| AccountCallPermissionsData {
-                        primary_did,
-                        secondary_key: Some(sk),
-                    });
+                // `who` is the primary key.
+                return Some(AccountCallPermissionsData {
```

Comment:
I think this can be returned from the `else` branch of
```
                if target_did == key_did {
                    // In this case, we check the permissions of `who` in `target_did` as it is a direct call
                    if who != &target_did_record.primary_key {
                         ...
                    } else {
                         // here
                    }
```
and in that case `return`s could be removed, so that we return the final value in each conditional branch.